mpqp/core/circuit.py

```diff
@@ -1045,16 +1052,78 @@ def to_qasm2(self) -> str:
             measure q[1] -> c[1];
 
         """
+        # TODO: put back this example when we figure out why the qasm2 export is different on docker/github actions
+        # >>> c2 = QCircuit([CustomGate(UnitaryMatrix(np.array([[0,1,0,0],[1,0,0,0],[0,0,0,1],[0,0,1,0]])),[1,2])])
+        # >>> print(c2.to_qasm2())
+        # OPENQASM 2.0;
+        # include "qelib1.inc";
+        # qreg q[3];
+        # u(0,pi/2,-pi/2) q[1];
+        # u(pi,-pi/2,pi/2) q[2];
+
+        from qiskit import qasm2, transpile, QuantumCircuit
+        from qiskit.circuit import CircuitInstruction
+
+        def replace_custom_gate(
+            custom_unitary: CircuitInstruction, nb_qubits: int
+        ) -> tuple[QuantumCircuit, float]:
+            """
+            Decompose and replace the (custom) qiskit unitary given in parameter by a ``QuantumCircuit`` composed of
+            ``U`` and ``CX`` gates, and return the global phase (related to usage of ``u`` in OpenQASM2) of the circuit.
+            Args:
+                custom_unitary: Qiskit CircuitInstruction containing the custom unitary operator.
+                nb_qubits: Number of qubits of the circuit from which the unitary instruction was taken.
+
+            Returns:
+                QuantumCircuit containing the decomposition of the unitary in terms of native gates ``U`` and ``CX``,
+                and the global phase used for final correction of the statevector.
+            """
+            from qiskit.exceptions import QiskitError
+
+            transpilation_circuit = QuantumCircuit(nb_qubits)
+            transpilation_circuit.append(custom_unitary)
+            try:
+                transpiled = transpile(transpilation_circuit, basis_gates=['u', 'cx'])
+            except QiskitError as e:
+                # if the error is arising from TwoQubitWeylDecomposition, we replace the matrix by the closest unitary
+                if "TwoQubitWeylDecomposition" in str(e):
+                    custom_unitary.operation.params[0] = closest_unitary(
+                        custom_unitary.operation.params[0]
+                    )
+                    transpiled = transpile(
+                        transpilation_circuit, basis_gates=['u', 'cx']
+                    )
+                else:
+                    raise e
+            return transpiled, transpiled.global_phase
+
         qiskit_circ = self.subs({}, remove_symbolic=True).to_other_language(
             Language.QISKIT
         )
         if TYPE_CHECKING:
             assert isinstance(qiskit_circ, QuantumCircuit)
 
-        from qiskit import qasm2
+        qiskit_circ.reverse_bits()
+
+        if all([not isinstance(inst, CustomGate) for inst in self.instructions]):
```

Comment:
why is this needed ? as I see it the following loop will cover this case, and it doesn't seem to save a lot of time

Reply:
the key point is that the two loops are not nested but one after the other, so it's `n+n` and not `n*n`, so I'd say it's not a big deal, but if you prefer to keep the shortcut I'm fine with it

Reply:
I introduced a boolean variable to check if wether we encountered a CustomGate in the circuit or not, and then dump the original or new qiskit circuit.

Reply:
I removed this variable (with your permission) because it doesn't really has any role in the result